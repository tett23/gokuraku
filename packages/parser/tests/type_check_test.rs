extern crate parser;

use insta::{assert_debug_snapshot, glob};
use std::fs;

#[test]
fn type_check_test() {
    glob!("../fixtures/pds", "*.pds", |path| {
        dbg!(&path);
        let input = fs::read_to_string(path).unwrap();
        let ast = parser::prose_down_script_parse(&input).unwrap();
        let ir1 = parser::ir::ir1::transform1(&ast);
        let ir2 = parser::ir::ir2::transform2(ir1, ast);

        assert_debug_snapshot!(ir2);
    });
}
