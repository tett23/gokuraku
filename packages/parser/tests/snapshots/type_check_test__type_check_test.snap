---
source: packages/parser/tests/type_check_test.rs
expression: ir
input_file: packages/parser/fixtures/pds/6.pds
---
Ok(
    SymbolTable {
        type_symbols: {
            Ident(
                "id",
            ): TypeAbstruction {
                arg: None,
                expr: Value(
                    Bottom,
                ),
            },
            Ident(
                "ternary",
            ): TypeAbstruction {
                arg: Some(
                    Value(
                        Bottom,
                    ),
                ),
                expr: Abstruction(
                    TypeAbstruction {
                        arg: Some(
                            Value(
                                Bottom,
                            ),
                        ),
                        expr: Abstruction(
                            TypeAbstruction {
                                arg: Some(
                                    Value(
                                        Bottom,
                                    ),
                                ),
                                expr: Value(
                                    Bottom,
                                ),
                            },
                        ),
                    },
                ),
            },
            Ident(
                "unary",
            ): TypeAbstruction {
                arg: Some(
                    Value(
                        Bottom,
                    ),
                ),
                expr: Value(
                    Bottom,
                ),
            },
            Ident(
                "binary",
            ): TypeAbstruction {
                arg: Some(
                    Value(
                        Bottom,
                    ),
                ),
                expr: Abstruction(
                    TypeAbstruction {
                        arg: Some(
                            Value(
                                Bottom,
                            ),
                        ),
                        expr: Value(
                            Bottom,
                        ),
                    },
                ),
            },
        },
        data_lattice: BoundedDataLattice {
            top: [
                DataLattice {
                    elements: [
                        DataElement {
                            ident: ExistsIdent(
                                "Unit",
                            ),
                            value: Unit,
                        },
                    ],
                    union_elements: None,
                    intersection_elements: None,
                },
                DataLattice {
                    elements: [
                        DataElement {
                            ident: ExistsIdent(
                                "Int",
                            ),
                            value: Unit,
                        },
                    ],
                    union_elements: None,
                    intersection_elements: None,
                },
            ],
            botom: [
                DataLattice {
                    elements: [
                        DataElement {
                            ident: ExistsIdent(
                                "Unit",
                            ),
                            value: Unit,
                        },
                    ],
                    union_elements: None,
                    intersection_elements: None,
                },
                DataLattice {
                    elements: [
                        DataElement {
                            ident: ExistsIdent(
                                "Int",
                            ),
                            value: Unit,
                        },
                    ],
                    union_elements: None,
                    intersection_elements: None,
                },
            ],
        },
    },
)
